adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -744,41 +978,244 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] {
+        // isn't important. we also avoid presorting the data because
+        // we end up shuffling the data around less. We simply perform a
+        // partition then sort by ReferenceRegion
+        flattenRddByRegions()
+          .mapPartitions(iter => {
+            // this is how we determine the destination partition number
+            iter.map(f => {
+              // inside the map over the partition we do a map over the partition map
+              // this implementation is in lieu of for...yield
+              destinationPartitionMap.zipWithIndex.map(partition => {
+                // if there is overlap between the partition map and the current value
+                if (f._1.covers(partition._1._1) || f._1.covers(partition._1._2) ||
+                  (f._1.compareTo(partition._1._1) >= 0 && f._1.compareTo(partition._1._2) <= 0)) {
+                  // we use Options here so scala doesn't lose typing information
+                  Some((partition._2, f))
+                } else if (partition._2 == destinationPartitionMap.length - 1 &&
+                  f._1.compareTo(partition._1._2) > 0) {
+                  Some((partition._2, f))
+                  // we do this so we don't filter out any data at the end or beginning
+                  // we don't want to lose data before the join, and this prevents that
+                } else if ((partition._2 == destinationPartitionMap.length - 1 && f._1.compareTo(partition._1._2) > 0) ||
+                  (partition._2 == 0 && f._1.compareTo(partition._1._1) < 0)) {
+                  Some((partition._2, f))
+                } else None
+              })
+            })
+          }, preservesPartitioning = true)
+          // finally we filter out the None values and flatten down
+          // the result is an RDD[(Int, (ReferenceRegion, T))].
+          // the Int is the destination partition index
+          .flatMap(f => f.filter(_.isDefined).map(g => g.get))
+          .map(f => ((f._2._1, f._1), f._2))
+          .repartitionAndSortWithinPartitions(
+            new GenomicPositionRangePartitioner(numPartitions, destinationPartitionMap))
+          .map(_._2)
+      }
+    // here we get the new partition map
+    // for now, we can't use the left RDD's partition map because
+    // the bounds don't always include the values we copartitioned with.
+    // in other words, joins allow data that starts before a given
+    // ReferenceRegion to combine given they overlap.
+    // this means that we cannot use the bounds of the left RDD because
+    // they may not necessarily be the bounds of the right RDD
+    val newPartitionMap = finalPartitionedRDD.mapPartitions(iter => {
+      getPartitionMapBoundsFromRdd(iter)
+    }, preservesPartitioning = true).collect
+
+    replaceRdd(finalPartitionedRDD.values, Some(newPartitionMap))
+  }
+
+  /**
+   * Gets the partition bounds from a ReferenceRegion keyed Iterator
```

Comment:
This comment doesn't match the method name, are the bounds coming from an iterator or an RDD?